statsmodels/tsa/holtwinters.py

```diff
@@ -0,0 +1,472 @@
+    damped : bool
+        Should the trend component be damped.
+    seasonal : str {"add", "mul", None}
+        Type of seasonal component.
+    phi : float
+        The phi value of the damped method, if the value is
+        set then this value will be used as the value.
+    optimised : bool
+        Should the values that have not beem set above be optimised automatically?
+    boxcoxed : bool
+        Should the boxcox tranform be applied to the data first?
+    remove_bias : bool
+        Should the bias be removed from the fcast and fitted values before being
+        returned?
+    try_harder : bool
+        Should the opptimser try harder to find optimal values?\
+
+    Returns
+    -------
+    results : namedtuple
+        A namedtuple with fcast, fitted and other attributes
+
+    Notes
+    -----
+    This is a full implementation of the holt winters exponential smoothing as 
+    per [1]. This includes all the unsatble methods as well as the stable methods.
+    The implmentaion of the library follows as per the R library as much as possible
+    whilest still being pythonic.
+    
+    References
+    ----------
+    [1] Hyndman, Rob J., and George Athanasopoulos. Forecasting: principles and practice. OTexts, 2014.
+    """
+    opt=None
+    trending = trend in ['mul','add']
+    seasoning = seasonal in ['mul','add']    
+    if damped and not trending:
+        raise NotImplementedError('Can only dampen the trend component')
+    phi = phi if damped else 1.0
+    N = len(data)
+    #if not isinstance(data, pd.Series):
+        #raise NotImplementedError('Only Pandas Series Supported')
+    if boxcoxed == True:
+        y, lam6da = boxcox(data)
+    elif boxcoxed == 'log':
+        lam6da = 0.0
+        y = boxcox(data, 0.0)
+    else:
+        if isinstance(data, pd.Series):
+            y = data.values.flatten()
+        else:
+            y = data.flatten()
+    y_alpha = np.zeros((N,))
+    y_gamma = np.zeros((N,))
+    l = np.zeros((N,))
+    b = np.zeros((N,))
+    if seasoning:
+        if (m is None or m == 0):
+            raise NotImplementedError('Unable to detect season automatically')
+    else:
+        m = 0
+    s = np.zeros((N+m-1,))
+    p = np.zeros(6+m)
+    max_seen = np.finfo(float).max
+    def holt_init(x, xi):
+        p[xi] = x
+        alpha,beta,_,l0,b0,phi = p[:6];
+        alphac = 1 - alpha
+        betac = 1 - beta
+        y_alpha[:] = alpha*y
+        l[:] = 0; b[:] = 0;
+        l[0] = l0; b[0] = b0
+        warnings.filterwarnings("error")
+        return alpha,beta,phi,alphac,betac
+    def holt__(x, xi):
+        alpha,beta,phi,alphac,betac = holt_init(x, xi)        
+        try:            
+            for i in range(1, N):
+                l[i] = (y_alpha[i-1]) + (alphac * (l[i-1]))
+            return sqeuclidean(l, y)
+        except RuntimeWarning:
+            return max_seen
+        finally:
+            warnings.filterwarnings("once")
+    def holt_mul_dam(x, xi):
```

Comment:
This patter of nested function and globals make unit testing really hard since it is difficult to directly test the smallest unit (e.g. a test hat holt_mul_dam works correctly).